### ISSUE TITLE
add support for supply chain smell "no code signature"

### ISSUE THREAD

**Issue by monperrus** [chains-project/dirty-waters#12]
add support for supply chain smell "no code signature", no release signature

for differential analysis: the equivalent would be "changed signing signatures"

**Comment by randomicecube** [comment]
@monperrus what's the idea here, exactly? If _any_ commit present in the commit history up until the release point isn't signed, we flag it?

My concern with flagging signatures (and this may be a dumb thought) is this: think about scenarios where you have a PR, which may have unsigned commits, and then a maintainer merges the PR; even if that maintainer signs the merge commit -- which should be what matters -- the other commits would be merged too (unless it's a squash), and that could lead to flagging code that, where it matters, _is signed_. I also don't think there's a good way to programmatically "ignore commits if they're coming from a PR which was merged".

Do you see where I'm getting at it/do you agree with my sentiment here? I may be overthinking

**Comment by monperrus** [comment]
There are commit signatures and release signatures.
Here, we refer to release signatures.
https://repo1.maven.org/maven2/fr/inria/gforge/spoon/spoon-core/11.1.1-beta-9/spoon-core-11.1.1-beta-9-jar-with-dependencies.jar.asc

**Comment by randomicecube** [comment]
@algomaster99 out of curiosity, do you know of any projects where testing for code signature would provide valuable results? Because atm my code at the [code-signature](https://github.com/chains-project/dirty-waters/tree/diogo/code-signature) branch doesn't find any missing code signatures in spoon nor in StoyanTc/yammer-metrics - v3.0.1, and I'd like to confirm whether the tool was correctly working or not

**Comment by algomaster99** [comment]
> testing for code signature would provide valuable results

What do you mean by valuable here?

In Java ecosystem, all maven releases on Central repository must confirm to [some requirements](https://central.sonatype.org/publish/release/#locate-and-examine-your-staging-repository). One of them is that all files that are pushed for deployment must be signed with gpg keys. However, there are [several repositories](https://mvnrepository.com/repos) where these requirements won't be enforced and there you may find some "no code signatures".

**Comment by randomicecube** [comment]
> What do you mean by valuable here?
I meant repos where, for one reason or another, it's known that, if we ran the tool, we'd find dependencies which aren't signed.

I didn't know that in the central repository it was a requirement to sign the code, that's nice to know! I'd say that this makes this check a bit less relevant then, but still, since packages might be in other repos which aren't the central one, it's probably worthwhile to keep it, right?

**Comment by algomaster99** [comment]
> might be in other repos which aren't the central one, it's probably worthwhile to keep it, right?

Exactly, that is my intuition as well. Although I have never come across such a repository. Let me as Hérve.
Also, @randomicecube I linked the wrong URL to requirements. I have updated it. Please check.

@hboutemy, do you know if [these requirements](https://central.sonatype.org/publish/requirements/) are only for `Central` repository in mvnrepository.com or it is for [all repositories](https://mvnrepository.com/repos)?

**Comment by hboutemy** [comment]
these requirements are for Maven Central only: other repositories may choose to reuse, but it's their choice to do or not to do (and implement checks if they don't have a centralized publication pipeline)

notice: mvnrepository is only a search engine, scraping as many repositories as they can, and consolidating indexes to provide consolidated search.

Each repository scraped by mvnrepository is managed by independent entities that do what they want with their content.

Maven Central has published rules and checks them, with most importantly namespace ownership https://central.sonatype.org/publish/requirements/coordinates/ and immutability https://central.sonatype.org/publish/requirements/immutability/

What do others do? I don't know: one consuming from them should ask them, because de-facto they trust them, and they trust them on any coordinates...
